components/omega/doc/design/Halo.md

```diff
@@ -0,0 +1,260 @@
+<!-- Halo Requirements and Design --------------------------------------------->
+
+# OMEGA Requirements and Design:
+
+## *Halo*
+
+## 1 Overview
+
+In a distributed memory parallel environment, it is necessary to regularly 
+exchange data on the interfaces between adjacent partitions of the mesh 
+(i.e. halo elements). These halo exchanges will be accomplished utilizing 
+the MPI library.
+
+## 2 Requirements
+
+### 2.1 Requirement: Compatible with all types of arrays
+
+Handle halo exchanges for arrays of each supported data type and dimensionality.
+
+### 2.2 Requirement: Exchange at all types of mesh elements
+
+Meshes are composed of three types of elements (cells, edges, and vertices) 
+at which values can be defined, we must be able to handle exchanges at each 
+type of element.
+
+### 2.3 Requirement: Variable halo depth
+
+The default minimum required halo depth for most of the algorithms we will 
+use is three cells. We should also be able to handle different halo depths.
+
+### 2.4 Requirement: Exchange lists
+
+Each MPI rank needs a send list and receive list for every neighboring rank and 
+each type of mesh element (cells, edges, and vertices) containing the local
+indices for each owned element to be packed and sent to neighboring ranks and 
+each element owned by neighboring ranks to be received and unpacked into local 
+field arrays. Exchange lists should be organized by halo layer.
+
+### 2.5 Requirement: Send and receive buffers
+
+Each MPI rank needs two arrays for every neighboring rank to serve as a send 
+buffer and a receive buffer to be passed to the MPI send and receive routines.
+Only one buffer needs to be communicated to each neighbor at a time, so the 
+buffers can be persistent and reused for each exchange to avoid allocating and 
+deallocating new buffer arrays for each exchange. The buffer size would 
+therefore need to be set to the largest necessary buffer length and only a 
+subset of the buffer array needs to be communicated for smaller exchanges. 
+
```

Comment:
I agree that this is desirable, but probably isn't required. It could be difficult to achieve when accounting for all the vertical levels and difficult decompositions, cases where the number of neighbors might be large and possibly managing communication from the device. So I think we should try but might need to change this to desired rather than required if we run into memory issues.